icefoganalytics/yhsi
# Issue 27: Change Tracking and Approval: Add Associations sub-form to Approvibility pathway

Fixes #27
Depends on:
- https://github.com/icefoganalytics/yhsi/pull/26

# User Story
See https://github.com/icefoganalytics/yhsi/issues/3
As an edit level user I can edit the /sites Associations form and it will create a site change request for admin approval.

# Implementation Details
- Upgraded PlaceEdit serialization to increase readability.
- The original PlaceEdit table has a slightly more complex format for the `FirstNationAssociationJSON` than I found to be required. I would need to write/run a migration if the old PlaceEdit records must to be kept.
```js
// new
[
	{
		"Id": 4612,
		"PlaceId": 1212,
		"FirstNationId": 13,
		"FirstNationAssociationType": 2,
		"Comments": ""
	},
	{
		"PlaceId": "1212",
		"FirstNationAssociationType": 1,
		"FirstNationId": 1,
		"Comments": "TEST"
	}
]

// old
[
	{
		"Id": 2913,
		"PlaceId": 2061,
		"FirstNationId": 13,
		"FirstNation": { "Id": 13, "Description": "Trondek Hwechin" },
		"FirstNationAssociationType": 2,
		"Comments": null
	}
]
```

# Screenshots
![image](https://user-images.githubusercontent.com/23045206/164055833-6cdcb0c9-4b83-4fbc-8963-ca7ce3018f71.png)

# Testing Instructions
1. Update your user roles to include `Site Admin` via the database console.
2. Go to the /sites table and select a site.
3. In the Associations form, make an edit and press save.
4. The site will refresh and your edits will be saved.
5.  Update your user roles so that they _do not include_ `Site Admin` or `Administrator` via the database console. Something like `Site Editor` would probably be best.
6. Go to the /sites table and select a site.
7. In the Associations form, make and edit (or two) and press save.
8. The page will refresh and all Associations fields will become readonly. The save button on the Associations form will be disabled with an appropriate tooltip.
9. Make yourself an `Site Admin` again and go to the /sites-change-requests table.
10. The new change request will be at the top of the list. Click on it to review.
11. Note the various features:
   - "accept/reject" toggle buttons
   - accept/reject all
   - save button is locked until all changes are approved/rejected
12. Reject and accept some changes and press Save.
13. Go back to the /sites table and check that you can now edit the site and that the changes have been applied or rejected correctly.
